tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_PrimitiveFromConservative.cpp

```diff
@@ -0,0 +1,212 @@
+Scalar<DataType> specific_enthalpy(
+    const Scalar<DataType>& rest_mass_density,
+    const Scalar<DataType>& specific_internal_energy,
+    const Scalar<DataType>& pressure) noexcept {
+  return Scalar<DataType>{1.0 + get(specific_internal_energy) +
+                          get(pressure) / get(rest_mass_density)};
+}
+
+template <typename DataType>
+Scalar<DataType> lorentz_factor(
+    const Scalar<DataType>& spatial_velocity_squared) noexcept {
+  return Scalar<DataType>{1.0 / sqrt(1.0 - get(spatial_velocity_squared))};
+}
+
+template <typename DataType>
+bool bad_velocity(const Scalar<DataType>& spatial_velocity_squared) noexcept {
+  const auto& vsq = get(spatial_velocity_squared);
+  for (size_t i = 0; i < get_size(vsq); ++i) {
+    if (get_element(vsq, i) > 1.0) {
+      return true;
+    }
+  }
+  return false;
+}
+
+template <typename DataType>
+bool bad_metric(const Scalar<DataType>& det_spatial_metric) noexcept {
+  const auto& gamma = get(det_spatial_metric);
+  for (size_t i = 0; i < get_size(gamma); ++i) {
+    if (get_element(gamma, i) < 0.0) {
+      return true;
+    }
+  }
+  return false;
+}
+
+template <typename DataType>
+bool fail_dominant_energy_condition(
+    const Scalar<DataType>& pressure, const Scalar<DataType>& rest_mass_density,
+    const Scalar<DataType>& specific_internal_energy) noexcept {
+  const auto& p = get(pressure);
+  const auto& rho = get(rest_mass_density);
+  const auto& epsilon = get(specific_internal_energy);
+  for (size_t i = 0; i < get_size(p); ++i) {
+    const double a = get_element(p, i) /
+                     (get_element(rho, i) * (1.0 + get_element(epsilon, i)));
+    if (a < 0.0 or a > 1.0) {
+      return true;
+    }
+  }
+  return false;
+}
+
+template <size_t Dim, typename DataType>
+void test_primitive_from_conservative(
+    const gsl::not_null<std::mt19937*> generator,
+    const gsl::not_null<std::uniform_real_distribution<>*> distribution,
+    const DataType& used_for_size) noexcept {
+  bool successful = false;
+
+  while (!successful) {
+    const auto expected_rest_mass_density =
+        make_with_random_values<Scalar<DataType>>(generator, distribution,
+                                                  used_for_size);
+    const auto expected_specific_internal_energy =
+        make_with_random_values<Scalar<DataType>>(generator, distribution,
+                                                  used_for_size);
+
+    const std::unique_ptr<EquationsOfState::EquationOfState<true, 2>>
+        equation_of_state =
+            std::make_unique<EquationsOfState::DarkEnergyFluid<true>>(0.5);
+    const auto expected_pressure = equation_of_state->pressure_from_density(
+        expected_rest_mass_density, expected_specific_internal_energy);
+
+    if (fail_dominant_energy_condition(expected_pressure,
+                                       expected_rest_mass_density,
+                                       expected_specific_internal_energy)) {
+      continue;  // try again
```

Comment:
Do we need to account for the possibility that this might always fail (so we never exit the loop?)

Reply:
at worst the test would time out then; I ran the test several thousand times and it never happened